File: ports/java/src/core/global/BcsString.java
package bcs.core.global;

public class BcsString {
    private final String value;

    public BcsString(String val) {
        this.value = (val != null) ? val : "";
    }

    public String toUpper() {
        return this.value.toUpperCase();
    }

    public String toLower() {
        return this.value.toLowerCase();
    }

    @Override
    public String toString() {
        return this.value;
    }

    public int length() {
        return this.value.length();
    }

    public boolean isEmpty() {
        return this.value.isEmpty();
    }

    public String substr(int index, int length) {
        if (index < 0 || index >= this.value.length()) return "";
        int end = Math.min(index + length, this.value.length());
        return this.value.substring(index, end);
    }

    public byte[] toUtf8() {
        return this.value.getBytes(java.nio.charset.StandardCharsets.UTF_8);
    }
}

File: ports/csharp/src/core/global/BcsString.cs
using System;

namespace BCS.Core.Global {
    public class BcsString {
        private string _value;

        public BcsString(string val) {
            _value = val ?? string.Empty;
        }

        public string ToUpper() {
            return _value.ToUpperInvariant();
        }

        public string ToLower() {
            return _value.ToLowerInvariant();
        }
<<<<<<< HEAD

=======

>>>>>>> origin/master
        public override string ToString() {
            return _value;
        }
    }
}

namespace BCS.Core.Global {
    public partial class BcsString {
        public int Length() {
            return _value.Length;
        }

        public bool IsEmpty() {
            return string.IsNullOrEmpty(_value);
        }

        public string Substr(int index, int length) {
            if (index < 0 || index >= _value.Length) return string.Empty;
            int len = System.Math.Min(length, _value.Length - index);
            return _value.Substring(index, len);
        }

        public byte[] ToUtf8() {
            return System.Text.Encoding.UTF8.GetBytes(_value);
        }
    }
}
